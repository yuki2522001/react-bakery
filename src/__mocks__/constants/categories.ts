import { CategoryProps } from "@common-types/category";

const CATEGORY_MOCKING_LIST: CategoryProps[] = [
  {
    id: "1",
    name: "Cake"
  },
  {
    id: "2",
    name: "Breads"
  }
]

export { CATEGORY_MOCKING_LIST };
