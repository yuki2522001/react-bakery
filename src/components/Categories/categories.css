.categories {
  margin-top: 50px;
  text-align: center;
}

.categories__title {
  text-transform: uppercase;
  font-weight: bold;
}

.categories__list {
  display: flex;
  justify-content: center;
  margin: 30px 0 40px 0;
  padding: 0;
}

.categories__item {
  list-style: none;
  cursor: pointer;
  margin: 0 30px;
  position: relative;
  color: var(--text-secondary);
}

.categories__item:hover {
  color: var(--text-primary);
}

.active {
  color: var(--text-primary);
}

.categories__item > a {
  color: var(--text-third);
  text-decoration: none;
}