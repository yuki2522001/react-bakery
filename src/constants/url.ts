const PRODUCTS_URL = `${process.env.VITE_BASE_URL}/products`;
const CATEGORIES_URL = `${process.env.VITE_BASE_URL}/categories`;

export { PRODUCTS_URL, CATEGORIES_URL };