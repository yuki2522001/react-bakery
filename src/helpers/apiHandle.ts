import axios, { AxiosResponse } from "axios";

// Get data form server
const getData = async <T>(url: string): Promise<T> => {
  const res = await axios.get<T>(url);
  return res.data;
};

// Post new data to server
const create = async <T>(url: string, newData: T):Promise<AxiosResponse<T>> => {
  const res = await axios.post(url, newData);
  return res;
}

// Send a changed data to server using method put to updating the data
const update = async <T>(id: string, dataEdit: T):Promise<AxiosResponse<T>> => {
  const res = await axios.put(id, {...dataEdit});
  return res;
}

// Delete data
const remove = async <T>(id: string): Promise<AxiosResponse<T>> => {
  const res = await axios.delete(id);
  return res;
}

export {
  getData,
  create,
  update,
  remove
}