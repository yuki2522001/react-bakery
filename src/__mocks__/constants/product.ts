import { Product } from "@common-types/product";
import url from "../../assets/images/products/black-forest-cake.jpg";

const PRODUCT_MOCKING: Product = {
  id: "1",
  categoryId: "1651924356334",
  name: "Cheese cake",
  price: 20,
  quantity: 5,
  description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
  images: [
    url
  ]
}

const PRODUCT_MOCKING_LIST: Product[] = [
  {
    id: "1",
    categoryId: "1651924356334",
    name: "Cheese cake",
    price: 20,
    quantity: 5,
    description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
    images: [
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg"
    ]
  },
  {
    id: "2",
    categoryId: "1651924356334",
    name: "Amaretti Cookie",
    price: 10,
    quantity: 5,
    description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
    images: [
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg"
    ]
  }
]

export { PRODUCT_MOCKING, PRODUCT_MOCKING_LIST };
