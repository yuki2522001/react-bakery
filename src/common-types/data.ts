import { Product } from "./product";

export enum Action {
  GetProductSuccess = "GET_PRODUCT_SUCCESS",
  CreateProductsSuccess = "CREATE_PRODUCT_SUCCESS",
  DeleteProductSuccess = "DELETE_PRODUCT_SUCCESS",
  UpdateProductSuccess = "UPDATE_PRODUCT_SUCCESS",
}

export interface DataState {
  products: Product[];
}

export type DataPayload = Product[] | string | Product

export interface DataAction {
  action: Action;
  payload: DataPayload
}
