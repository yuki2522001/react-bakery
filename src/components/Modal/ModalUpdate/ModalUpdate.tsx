import useSWR from "swr";
import React, { ChangeEvent, useCallback, useContext, useState } from "react";
import toast from "react-hot-toast";
import getBase64 from "@helpers/getBase64";
import InputValue from "@components/Input/InputValue/InputValue";
import Button from "@components/common/Button/Button/Button";
import { FieldName, FormProps } from "@common-types/form";
import { Product } from "@common-types/product";
import { RULES } from "@constants/rules";
import { validate } from "@helpers/validate";
import { CategoryProps } from "@common-types/category";
import { SUCCESS_MSG } from "@constants/message";
import { getData, update } from "@helpers/apiHandle";
import { setFieldsValue } from "@helpers/fieldHandle";
import { CATEGORIES_URL, PRODUCTS_URL } from "@constants/url";
import { DataContext } from "@context/DataContext";
import { Action } from "@common-types/data";
import "../modal.css";

interface ModalUpdateProps {
  product: Product;
  hideModalUpdate: () => void;
  deleteImage: () => void;
  updateProductDetail: (product: Product) => void;
}

const ModalUpdate: React.FC<ModalUpdateProps> = ({
  product,
  hideModalUpdate,
  updateProductDetail,
}) => {
  // fetch data with useSWR
  const { data } = useSWR(CATEGORIES_URL, getData<CategoryProps[]>);
  const {dispatch} = useContext(DataContext);
  // create state to handle select file image
  const [selectedFile, setSelectedFile] = useState<string[]>([]);
  // create state to update product
  const [productEdit, setProductEdit] = useState<Product>(product);
  const updateProduct = async (id: string, productData: Product) => {
    const productEdit: Product = {
      id: productData.id,
      categoryId: productData.categoryId,
      name: productData.name,
      price: +productData.price,
      quantity: +productData.quantity,
      description: productData.description,
      images: productData.images,
    };
    try {
      const response = await update(`${PRODUCTS_URL}/${id}`, productEdit);
      if(response) {
        dispatch({
          action: Action.UpdateProductSuccess,
          payload: {...response.data}
        });
        updateProductDetail(response.data);
      }
      toast.success(SUCCESS_MSG.MESSAGE_UPDATE_PRODUCT);
    } catch (error: any) {
      toast.error(error.message);
    }
  };

  const [formValues, setFormValues] = useState<FormProps>({
    categoryId: {
      value: productEdit?.categoryId,
      rules: [RULES.REQUIRED],
      error: "",
    },
    name: {
      value: productEdit?.name,
      rules: [RULES.REQUIRED],
      error: "",
    },
    price: {
      value: productEdit?.price.toString(),
      rules: [RULES.REQUIRED, RULES.NUMBER, RULES.NEGATIVE],
      error: "",
    },
    quantity: {
      value: productEdit?.quantity.toString(),
      rules: [RULES.REQUIRED, RULES.NUMBER, RULES.NEGATIVE],
      error: "",
    },
    description: {
      value: productEdit?.description,
      rules: [RULES.REQUIRED],
      error: "",
    },
    images: {
      value: productEdit?.images.toString(),
      rules: [],
      error: "",
    }
  });

  // handle update product
  const handleUpdateProduct = (id: string, productEdit: Product) => {
    for (let i = 0; i < selectedFile.length; i++) {
      productEdit.images.push(selectedFile[i]);
    }

    // validate form
    setFormValues(validate(formValues));
    const temp = (Object.keys(formValues) as (keyof typeof formValues)[]).map(
      (fieldName) => {
        if (formValues[fieldName].error) {
          return false;
        }
      }
    );

    if (!temp.includes(false)) {
      setFormValues({...formValues});
      updateProduct(id, productEdit);
      hideModalUpdate();
    }
  };

  // handle change value
  const handleChange = useCallback((event: { target: { value: string; name: string } }) => {
    const value = event.target.value;
    const fieldName = event.target.name as FieldName;
    setProductEdit({ ...productEdit, [fieldName]: value });
    setFormValues(setFieldsValue(formValues, value, fieldName));
  }, [productEdit]);

  // handle change image
  const imageChange = useCallback(async (event: ChangeEvent<HTMLInputElement>) => {
    const files = event.target.files;
    const newFiles = [];
    if (files) {
      for (let i = 0; i < files.length; i++) {
        const imageSrc = await getBase64(files[i]);
        newFiles.push(imageSrc)
      }
    }
    setSelectedFile([...selectedFile, ...newFiles]);
  }, [selectedFile]);

  // handle delete image
  const handleDeleteImage = useCallback((event: { target: EventTarget }) => {
    const target = event.target as Element;
    const indexOfArr = productEdit.images.findIndex(
      (item: string) => item == (target as HTMLInputElement).dataset.id
    );
    const indexOf = selectedFile.findIndex(
      (item: string) => item == (target as HTMLInputElement).dataset.id
    );
    selectedFile.splice(indexOf, 1);
    productEdit.images.splice(indexOfArr, 1);
    setSelectedFile([...selectedFile]);
  },[selectedFile]);

  return (
    <div data-testid="modal-update" className="modal-update">
      <div className="modal-dialog-modalUpdate">
        <div className="modal-content">
          <div className="modal-header">
            <h4 className="modal-heading" id="productModalLabel">
              Edit Product Information
            </h4>
          </div>
          <div className="modal-body">
            <div className="form-control">
              <label htmlFor="">Product name: </label>
              <InputValue
                type="text"
                name="name"
                value={productEdit?.name}
                onChange={handleChange}
              />
              <small className="form__error">
                {formValues.name.error}
              </small>
            </div>
            <div className="form-control">
              <label htmlFor="">Description: </label>
              <textarea
                data-testid="change-value"
                className="input__text"
                cols={30}
                rows={5}
                value={productEdit?.description}
                name="description"
                onChange={handleChange}
              ></textarea>
              <small className="form__error">
                {formValues.description.error}
              </small>
            </div>
            <div className="form-control">
              <label htmlFor="">Categories: </label>
              <select
                className="form__select"
                name="categoryId"
                onChange={handleChange}
              >
                {data?.map(({ id, name }: CategoryProps, index: number) => (
                  <option
                    key={index}
                    value={id}
                  >
                    {name}
                  </option>
                ))}
              </select>
              <small className="form__error">
                {formValues.categoryId.error}
              </small>
            </div>
            <div id="form__number" className="form-control">
              <div className="form-control">
                <label htmlFor="">Price: </label>
                <InputValue
                  className="input__number"
                  type="number"
                  min={0}
                  value={productEdit?.price}
                  name="price"
                  onChange={handleChange}
                />
                <small className="form__error">
                  {formValues.price.error}
                </small>
              </div>
              <div className="form-control">
                <label htmlFor="">Quantity: </label>
                <InputValue
                  className="input__number"
                  type="number"
                  min={0}
                  value={productEdit?.quantity}
                  name="quantity"
                  onChange={handleChange}
                />
                <small className="form__error">
                  {formValues.quantity.error}
                </small>
              </div>
            </div>
            <div className="form-control">
              <div className="form__img--list">
                {productEdit?.images?.map((img: string, index: number) => (
                  <img
                    data-testid="delete-image"
                    key={index}
                    className="form__img"
                    data-id={img}
                    src={img}
                    onClick={handleDeleteImage}
                  />
                ))}
                {selectedFile.length > 0 &&
                  selectedFile.map((src, key: number) => (
                    <img
                      data-testid="after-change-file"
                      key={key}
                      className="form__img"
                      data-id={src}
                      src={src}
                      onClick={handleDeleteImage}
                    />
                  ))}
                <input
                  data-testid="change-file"
                  className="form__input--img"
                  type="file"
                  id="file"
                  multiple
                  name="images"
                  onChange={imageChange}
                />
              </div>
            </div>
          </div>
          <div className="modal-footer-modalUpdate">
            <Button
              type="warning"
              onClick={hideModalUpdate}
              text="Cancel"
            />
            <Button
              text="Submit"
              type="success"
              onClick={() => handleUpdateProduct(product.id, productEdit)}
            />
          </div>
        </div>
      </div>
    </div>
  );
};

export default ModalUpdate;
