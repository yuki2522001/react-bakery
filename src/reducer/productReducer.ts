import { Action, DataAction, DataState } from "@common-types/data";
import { Product } from "@common-types/product";

const productReducer = (state: DataState, actions: DataAction): DataState => {
  const { action, payload } = actions;
  switch (action) {
    case Action.GetProductSuccess: {
      return { ...state, products: payload as Product[] };
    }
    case Action.CreateProductsSuccess: {
      return { ...state.products, products: state.products.concat(payload as Product) };
    }
    case Action.DeleteProductSuccess: {
      const deleteProduct = state.products.filter(item => item.id != payload)
      return {...state, products: deleteProduct};
    }
    case Action.UpdateProductSuccess: {
      state.products.findIndex(item => item.id == payload);
      return {...state};
    }
    default: {
      return { ...state };
    }
  }
};

export { productReducer };
