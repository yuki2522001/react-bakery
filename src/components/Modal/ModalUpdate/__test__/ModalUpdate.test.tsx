import user from "@testing-library/user-event";
import mockAxios from "@__mocks__/axios";
import ModalUpdate from "../ModalUpdate";
import "@testing-library/jest-dom";
import { useState } from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { CATEGORIES_URL, PRODUCTS_URL } from "@constants/url";
import { CATEGORY_MOCKING_LIST } from "@__mocks__/constants/categories";
import { getData, update } from "@helpers/apiHandle";
import {
  PRODUCT_MOCKING,
  PRODUCT_MOCKING_LIST,
} from "@__mocks__/constants/product";
import { Action, DataState } from "@common-types/data";
import { productReducer } from "@reducer/productReducer";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

describe("Modal update component", () => {
  const hideModalUpdate = jest.fn();
  const updateProductDetail = jest.fn();
  const deleteImages = jest.fn();
  const someValues = [{ name: "cheese pocket" }];
  const setup = () => {
    const utils = render(
      <ModalUpdate
        product={PRODUCT_MOCKING}
        hideModalUpdate={() => {}}
        deleteImage={() => {}}
        updateProductDetail={() => {}}
      />
    );
    const input = utils.getByTestId("change-value") as HTMLInputElement;
    return {
      input,
      ...utils,
    };
  };

  beforeEach(() => {
    (useState as jest.Mock).mockImplementation(
      jest.requireActual("react").useState
    );
  });

  afterEach(() => {
    mockAxios.reset();
  });

  test("should change value when onChang input", () => {
    const { input } = setup();
    fireEvent.change(input, { target: { value: "Cheese pocket" } });
    expect(input.value).toBe("Cheese pocket");
  });

  test("should change file image when onChang input", async () => {
    const { getByTestId, queryByTestId } = render(
      <ModalUpdate
        product={PRODUCT_MOCKING}
        hideModalUpdate={() => {}}
        deleteImage={() => {}}
        updateProductDetail={() => {}}
      />
    );
    const str = JSON.stringify(someValues);
    const blob = new Blob([str]);
    const file = new File([blob], "db.json", {
      type: "application/JSON",
    });
    File.prototype.text = jest.fn().mockResolvedValueOnce(str);
    const input = getByTestId("change-file");
    user.upload(input, file);
    await waitFor(() =>
      expect(queryByTestId("after-change-file")).toBeTruthy()
    );
  });

  test("should delete image when click image", () => {
    const { getByTestId } = render(
      <ModalUpdate
        product={PRODUCT_MOCKING}
        hideModalUpdate={() => {}}
        deleteImage={deleteImages()}
        updateProductDetail={() => {}}
      />
    );
    const deleteImage = getByTestId("delete-image");
    fireEvent.click(deleteImage);
    expect(deleteImages).toHaveBeenCalled();
  });

  test("get categories item should call", async () => {
    mockAxios.get.mockResolvedValueOnce({ data: CATEGORY_MOCKING_LIST });
    const result = await getData(CATEGORIES_URL);
    expect(mockAxios.get).toHaveBeenCalledWith(CATEGORIES_URL);
    expect(result).toEqual(CATEGORY_MOCKING_LIST);
  });

  test("update product item should call", async () => {
    const PRODUCT_URL_CALL = PRODUCTS_URL + "/1";
    mockAxios.put.mockResolvedValueOnce(PRODUCT_MOCKING);
    const result = await update(PRODUCT_URL_CALL, PRODUCT_MOCKING);
    expect(mockAxios.put).toHaveBeenCalledWith(
      PRODUCT_URL_CALL,
      PRODUCT_MOCKING
    );
    expect(result).toEqual(PRODUCT_MOCKING);
  });

  test("should render modal update component", () => {
    const { getByTestId } = render(
      <ModalUpdate
        product={PRODUCT_MOCKING}
        hideModalUpdate={() => {}}
        deleteImage={() => {}}
        updateProductDetail={() => {}}
      />
    );
    expect(getByTestId("modal-update")).toBeInTheDocument();
  });

  test("should hide modal update when click Cancel", () => {
    render(
      <ModalUpdate
        product={PRODUCT_MOCKING}
        hideModalUpdate={hideModalUpdate}
        deleteImage={() => {}}
        updateProductDetail={() => {}}
      />
    );
    const hideModal = screen.getByText("Cancel");
    fireEvent.click(hideModal);
    expect(hideModalUpdate).toHaveBeenCalled();
  });

  test("should update product when click Submit", () => {
    render(
      <ModalUpdate
        product={PRODUCT_MOCKING}
        hideModalUpdate={updateProductDetail}
        deleteImage={() => {}}
        updateProductDetail={() => {}}
      />
    );
    const btnSubmit = screen.getByText("Submit");
    fireEvent.click(btnSubmit);
    expect(updateProductDetail).toHaveBeenCalled();
  });

  test("should update state when dispatch action UpdateProductSuccess", () => {
    const initialState: DataState = {
      products: PRODUCT_MOCKING_LIST,
    };
    const updateProduct = {
      action: Action.UpdateProductSuccess,
      payload: PRODUCT_MOCKING_LIST,
    };
    const updatedState = productReducer(initialState, updateProduct);
    expect(updatedState).toEqual(updatedState);
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <ModalUpdate
        product={PRODUCT_MOCKING}
        hideModalUpdate={() => {}}
        deleteImage={() => {}}
        updateProductDetail={() => {}}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
