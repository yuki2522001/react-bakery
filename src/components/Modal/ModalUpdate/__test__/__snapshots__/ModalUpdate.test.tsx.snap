// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal update component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="modal-update"
    data-testid="modal-update"
  >
    <div
      class="modal-dialog-modalUpdate"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <h4
            class="modal-heading"
            id="productModalLabel"
          >
            Edit Product Information
          </h4>
        </div>
        <div
          class="modal-body"
        >
          <div
            class="form-control"
          >
            <label
              for=""
            >
              Product name: 
            </label>
            <input
              name="name"
              type="text"
              value="Cheese cake"
            />
            <small
              class="form__error"
            />
          </div>
          <div
            class="form-control"
          >
            <label
              for=""
            >
              Description: 
            </label>
            <textarea
              class="input__text"
              cols="30"
              data-testid="change-value"
              name="description"
              rows="5"
            >
              Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.
            </textarea>
            <small
              class="form__error"
            />
          </div>
          <div
            class="form-control"
          >
            <label
              for=""
            >
              Categories: 
            </label>
            <select
              class="form__select"
              name="categoryId"
            />
            <small
              class="form__error"
            />
          </div>
          <div
            class="form-control"
            id="form__number"
          >
            <div
              class="form-control"
            >
              <label
                for=""
              >
                Price: 
              </label>
              <input
                class="input__number"
                min="0"
                name="price"
                type="number"
                value="20"
              />
              <small
                class="form__error"
              />
            </div>
            <div
              class="form-control"
            >
              <label
                for=""
              >
                Quantity: 
              </label>
              <input
                class="input__number"
                min="0"
                name="quantity"
                type="number"
                value="5"
              />
              <small
                class="form__error"
              />
            </div>
          </div>
          <div
            class="form-control"
          >
            <div
              class="form__img--list"
            >
              <input
                class="form__input--img"
                data-testid="change-file"
                id="file"
                multiple=""
                name="images"
                type="file"
              />
            </div>
          </div>
        </div>
        <div
          class="modal-footer-modalUpdate"
        >
          <button
            class="btn btn__warning"
          >
            Cancel
          </button>
          <button
            class="btn btn__success"
          >
            Submit
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
